src/Xrefcheck/Verify.hs

```diff
@@ -426,94 +454,103 @@ verifyReference
-
-      atomicModifyIORef' progressRef $ \VerifyProgress{..} ->
-        ( if isExternal $ locationType rLink
-          then VerifyProgress{ vrExternal =
-            let vrExternalAdvanced = moveProgress vrExternal
-            in if toRetry
-               then case pTaskTimestamp vrExternal of
-                      Just (TaskTimestamp ttc start)
-                        | currentRetryAfter +:+ now <= ttc +:+ start -> vrExternalAdvanced
-                      _ -> setTaskTimestamp currentRetryAfter now vrExternalAdvanced
-               else vrExternalAdvanced, .. }
-          else VerifyProgress{ vrLocal = moveProgress vrLocal, .. }
-        , ()
-        )
-      if toRetry
-      then do
-        threadDelay currentRetryAfter
-        retryVerification (numberOfRetries + 1) resIO
-      else return $ fmap (WithReferenceLoc fileWithReference ref) res
-
-    alterOverallProgress
+      case res of
+        Nothing -> do
+          let moveProgressOnSuccess =
+                incProgress
+                  . if errorsOccured rc
+                    then decProgressFixableErrors
+                    else id
+          modifyProgressRef now Nothing moveProgressOnSuccess
+          pure res
+        Just err ->
+          case extractRetryAfterInfo err of
+            Nothing -> do
+              let moveProgressOnUnfixable =
+                    incProgress
+                      . if errorsOccured rc
+                        then fixableToUnfixable
+                        else incProgressUnfixableErrors
+              modifyProgressRef now Nothing moveProgressOnUnfixable
+              pure res
+            Just (mbCurrentRetryAfter, counterModifier, fetchCounter, fetchCorrespondingMaximum) -> do
+              let toSeconds = \case
+                    Seconds s -> s
+                    -- Calculates the seconds left until @Retry-After@ date.
+                    -- Defaults to 0 if the date has already passed.
+                    Date date | utcTimeToTimeSecond date >= now -> utcTimeToTimeSecond date -:- now
+                    _ -> sec 0
+
+              let toRetry = rcTooManyRequestsRetries < ncMaxRetries cNetworking
+                              && rcTimeoutRetries < ncMaxTimeoutRetries cNetworking
+                  currentRetryAfter = fromMaybe (ncDefaultRetryAfter cNetworking) $
+                    fmap toSeconds mbCurrentRetryAfter
+
+              let moveProgress =
+                    alterProgressErrorsOnFixable (fetchCorrespondingMaximum cNetworking) (fetchCounter rc)
+                      . if not toRetry
+                        then incProgress
+                        else id
+              modifyProgressRef now (Just currentRetryAfter) moveProgress
+              if toRetry
+              then do
+                threadDelay currentRetryAfter
+                retryVerification
+                  (counterModifier rc)
+                  resIO
+              else pure res
```

Comment:
I love this refactoring :+1:   Having these `case`s simplifies the downstream conditions.

I would now also try to split this canvas of code, maybe by extracting the part not related to progress update to a separate function - if this turns out to be viable.

And a note on the future: it would be worth putting this refactoring to a separate commit before adding the timeouts handling, this would simplify the review. But for sure, no need to extract this now, how it currently is - is fine too.

Reply:
> I would now also try to split this canvas of code, maybe by extracting the part not related to progress update to a separate function - if this turns out to be viable.

It seems that this function is already OK:+1: now. If the decision is up to me, let's leave it as is.

Reply:
Okay, sounds nice :+1: